features/329-identity-apis/api.md

```diff
@@ -0,0 +1,363 @@
+# Design Summary
+
+## SDK API
+```typescript
+type Metadata = { [key: string]: unknown };
+
+interface User {
+    id: string;
+    createdAt: Date;
+    lastLoginAt: Date;
+    isVerified: boolean;
+    isDisabled: boolean;
+    metadata: Metadata;
+
+    identity: Identity;
+}
+
+interface PasswordIdentity {
+    id: string;
+    type: 'password';
+    loginIDKey: string;
+    loginID: string;
+    realm: string;
+    claims: {
+        email?: string;
+        phone?: string;
+        // or other standard keys in future
+    };
+}
+interface OAuthIdentity {
+    id: string;
+    type: 'oauth';
+    providerID: string;
+    providerUserID: string;
+    rawProfile: object;
+    claims: {
+        email?: string;
+        // or other standard keys in future
+    };
+}
+interface CustomTokenIdentity {
+    id: string;
+    type: 'custom';
+    providerUserID: string;
+    rawProfile: object;
+    claims: {
+        // or other standard keys in future
+    };
+}
+
+// all identity types have common fields: id, type and metadata.
+type Identity = PasswordIdentity | OAuthIdentity | CustomTokenIdentity;
+
+async function listIdentities(): Promise<Identity[]>;
+
+async function addLoginID(loginID: { [key: string]: string }, realm?: string): Promise<void>;
+async function removeLoginID(loginID: string, realm?: string): Promise<void>;
+async function updateLoginID(oldLoginID: string, newLoginID: { [key: string]: string }): Promise<User>;
+
+async function changePassword(newPassword: string, oldPassword?: string): Promise<User>;
+
+```
+
+
+## Configuration
+`disableReauthForSecurity` (boolean; default `false`):
+Require re-authentication (i.e. access token must be issued recently) for
+security-critical operations:
+- add/remove login ID
+- link/unlink SSO provider (except auto-link when logging in with OAuth)
+- change password (if old password is not provided)
+
+`reauthIntervalSeconds` (integer; default 300, i.e. 5 minutes):
+The period after the access token is issued, that is considered authenticated
+for security-critical operations.
+
+`updateLoginIDEnabled` (boolean; default `false`):
```

Comment:
I think this should default to `true`? Any reason why we need to disable it actually?

Reason is allow changing of username seems a basic enough use case? Or there are special reason why this feature is problematic? (As it should be obvious that we don't want to have feature flag for users to enable/disable each and every features)

Reply:
This feature has some important implications: many users & developers expect the user name would not be changed, so it requires much consideration on UX (e.g. user profile link slug, display name impersonation).

Reply:
Agree we should have this option. Not sure if we should have default to true or false. Let's stick with it first than if you guys think false is better.